### ISSUE TITLE
In VAB/SPH: constant disk read, disk write, growing memory

### ISSUE THREAD

**Issue by hatalmasalma** [linuxgurugamer/ksp_toolbar#27]
Probably a problem with WASD Editor Camera or something else, didn't test with other mods.

- GameData\WasdEditorCamera\Textures\WASD-on-24.png is read ~100 times per second
- GameData\toolbar-settings.dat is saved ~500 times per second
- The game's memory is growing bigger and bigger

- See the attached NirSoft tool export: 1785 / 357 = 5. So the .dat is saved exactly 5 times while the icon is loaded 1 time.

Replication:
1. Clean KSP Install, put mods' folders into GameData
2. New Save
3. Close the welcoming messages (I chose that the ClickThroughBlocker follows the mouse instead of clicks)
4. Enter VAB
5. Click the orange wrench icon, set WASDEditorCamera from Stock to Blizzy's only
6. Maybe have to exit and reenter VAB, but the problem appears also after restarting the game, reverting from flight

I double checked that there isn't gamedata folder in gamedata folder.

Same effect with ModuleManager installed, since I noticed this with my heavily modded pack.

```
GameData:

000_ClickThroughBlocker\
000_Toolbar\
001_ToolbarControl\
Squad\
WasdEditorCamera\
toolbar-settings.dat
```

```
Test environment:

Clean KSP 1.11.1 install
Without DLCs (the same effect with DLCs)

Blizzy's Toolbar / ksp_toolbar      (from Toolbar-1.8.0-1.8.0.5.zip)
ToolbarControl                           (from ToolbarControl-1.8.0-0.1.9.4.zip)
Click Through Blocker                 (from ClickThroughBlocker-1.10.1-0.1.10.15.zip)
WASD Editor Camera                  (from WasdEditorCamera-1.10.1-0.7.3.zip)
```

```
Logs: https://gist.github.com/hatalmasalma/cd6887de8941d4e8b1a550e59fe722bf
When making these logs, the settings have been already changed, so I just started the game, loaded the save and the VAB, then exited the VAB, then Alt+F4.
```

```
NirSoft ProcessActivityView export (have been logging for a few secs)

==================================================
Index             : 1
Filename          : C:\Kerbal Space Program v1-11-1_2\GameData\WasdEditorCamera\Textures\WASD-on-24.png
Extension         : png
Open Count        : 357
Close Count       : 357
Failed Count      : 0
Read Count        : 357
Write Count       : 0
Read Bytes        : 357 000
Write Bytes       : 0
Last Handle       : 0x00000000000005a8
Last Error Code   : 
Opened By Module  : 
Opened By Address : 0x0000000000300101
==================================================

==================================================
Index             : 2
Filename          : C:\Kerbal Space Program v1-11-1_2\GameData\toolbar-settings.dat
Extension         : dat
Open Count        : 357
Close Count       : 357
Failed Count      : 0
Read Count        : 0
Write Count       : 1785
Read Bytes        : 0
Write Bytes       : 215 271
Last Handle       : 0x00000000000005a8
Last Error Code   : 
Opened By Module  : 
Opened By Address : 0x0000000000300101
==================================================

```

**Comment by linuxgurugamer** [comment]
Given that literally tens of thousands of people are using this, I really doubt it's a problem with the mod.
I've left it sitting in the editor for hours without seeing this problem.

I'll take a look, but you would be better served by posting in the threads on the forum, where other people can help as well.

**Comment by linuxgurugamer** [comment]
Install ZeroMiniAVC.  You still have MiniAVC installed, it's been broken for a few years now

**Comment by hatalmasalma** [comment]
Tried a few things, but they didn't work:
-Deleting miniavc.dll
-Installing ZeroMiniAVC
-Installing ZeroMiniAVC+ deleting miniavc.dll
-changing resolution
-renaming the constantly loaded png (the .dat was still being rewritten)

Finally I found the problem:

The problem was with WASDEditorCamera's icon, or something related to that. Lucky (and lazy) me only installed that mod when testing.

Workaround:
I simply disabled WASDEditorCamera's icon from Blizzy Toolbar (in the settings page of Blizzy Toolbar. Of course I will have to reassign it to the stock toolbar if I want to modify something, but it will be hidden for now.


Anyway, thanks for your previous modding work

**Comment by linuxgurugamer** [comment]
That is odd, I'll try it out.   Thanks for the  update.  I'm going to reopen this until I get a chance to look at it

**Comment by KEKKJ123** [comment]
I originally encountered this issue on my RP-1 save, and I just retested it in a clear KSP.
I think my situation is the same as the previous player. Here is the information.
- Symptoms: Memory usage continues to grow, about 3MB/s; CPU usage grow about 2-2.5%; and disk usage grow about 0.5MB/s(when it happened)
- Reproduction: Use ckan to install Toolbar and WasdEditorCamera, open KSP and create a new game to enter VAB, set WASDEditorCamera from Stock to Blizzy's only (or both)
- LOGs: Some Notices: error(in log) does not always appear. I'll keep testing, if I have free time. :(
```
# A clear KSP
[LOG 03:08:29.391] [ApplicationLauncher] OnSceneLoadedGUIReady: scene EDITOR ShouldBeVisible() True ShouldBeOnTop() False iIsPositionedAtTop False
[LOG 03:08:29.393] ScaleModList: listSize 41 maxListSize 1575
[LOG 03:08:29.394] ScaleModList: listSize 82 maxListSize 1575
[LOG 03:08:29.394] ScaleModList: listSize 123 maxListSize 1575
[LOG 03:08:29.395] ScaleModList: listSize 164 maxListSize 1575
[LOG 03:08:29.399] [UIApp] OnDestroy: Contracts
```
```
# In my RP-1 career
[LOG 02:56:28.729] [FAR v0.16.1.2]: Std dev for smoothing: 3 voxel total vol: 640.869630341885 filled vol: 111.2379091945
[LOG 02:56:28.797] [FAR v0.16.1.2]: Voxelization Time (ms): 640
[ERR 02:56:33.188] [KSPCommunityFixes]: No Persistent fields on object of type MovingAverage that is referenced in persistent field, adding as null to TypeCache.
[ERR 02:58:33.611] [Toolbar] [ERROR] button texture not found: WasdEditorCamera/Textures/WASD-24
[ERR 02:58:33.611] [Toolbar] [ERROR] Current Dir: E:\KSP\RP-1
[ERR 02:58:33.611] [Toolbar] [ERROR] dir: E:\KSP\RP-1\GameData\WasdEditorCamera\Textures
[ERR 02:58:33.611] [Toolbar] [ERROR] filePath: E:\KSP\RP-1\GameData\WasdEditorCamera\Textures\WASD-24.png
[ERR 02:58:33.623] [Toolbar] [ERROR] file: E:\KSP\RP-1\GameData\WasdEditorCamera\Textures\WASD-24.png
[ERR 02:58:33.623] [Toolbar] [ERROR] file: E:\KSP\RP-1\GameData\WasdEditorCamera\Textures\WASD-38.png
[ERR 02:58:33.623] [Toolbar] [ERROR] file: E:\KSP\RP-1\GameData\WasdEditorCamera\Textures\WASD-disabled-24.png
[ERR 02:58:33.623] [Toolbar] [ERROR] file: E:\KSP\RP-1\GameData\WasdEditorCamera\Textures\WASD-disabled-38.png
[ERR 02:58:33.623] [Toolbar] [ERROR] file: E:\KSP\RP-1\GameData\WasdEditorCamera\Textures\WASD-on-24.png
[ERR 02:58:33.623] [Toolbar] [ERROR] file: E:\KSP\RP-1\GameData\WasdEditorCamera\Textures\WASD-on-38.png
[LOG 02:58:42.175] ScaleModList: listSize 369 maxListSize 1172
[LOG 02:58:42.182] [Adjustable Mod Panel] Encountered a known mod: module WASD_NS, method wasdButton
[LOG 02:59:03.470] ScaleModList: listSize 369 maxListSize 1172
[LOG 02:59:09.484] ScaleModList: listSize 369 maxListSize 1172
```
In summary, it can be basically determined that the problem is with WasdEditorCamera and Toolbar.
If you need me to upload the complete log file, please let me know.
I apologize for my poor English and thank you for your effort!

**Comment by linuxgurugamer** [comment]
Well, I don't know what to say.  I've just tested this, multiple ways, and I don't see this issue.
Please supply a complete Player.log so I can take a look.

**Comment by KEKKJ123** [comment]
That‘s strange... but here is the log, if you need anything else please let me know, thank you very much!

[KSP.log](https://github.com/user-attachments/files/19524576/KSP.log)

**Comment by linuxgurugamer** [comment]
> That‘s strange... but here is the log, if you need anything else please let me know, thank you very much!
> 
> [KSP.log](https://github.com/user-attachments/files/19524576/KSP.log)

First, that's not a full log file.  Or, if it is, it certainly isn't showing those errors.  The only messages I saw about Wasd were normal, none of the messages you showed earlier

Second, I asked for the Player.log, not KSP.log.  Player.log is a bit more low-level, and shows things in a slightly different order.

**Comment by KEKKJ123** [comment]
Sorry, I misunderstood, here is the Player.log
[Player.log](https://github.com/user-attachments/files/19525261/Player.log)

**Comment by linuxgurugamer** [comment]
I found the problem, it's actually a bug in WASD, a fix will be released soon

**Comment by linuxgurugamer** [comment]
A bit more investigation showed that while it was being caused by WASD, it was a small bug in the Toolbarcontroller.
Updated and released

**Comment by KEKKJ123** [comment]
> I found the problem, it's actually a bug in WASD, a fix will be released soon

So, what kind of bug could have gone unreported for so long? I was a little curious.
anyway, thank you again for your patience and work!

**Comment by linuxgurugamer** [comment]
> > I found the problem, it's actually a bug in WASD, a fix will be released soon
> 
> So, what kind of bug could have gone unreported for so long? I was a little curious. anyway, thank you again for your patience and work!

Something that most people wouldn't have encountered, and those who did, didn't realize what it was. 
WASD updates the button every refresh, the bug was with the Blizzy toolbar code, where it loaded the button from disk every refresh.  AFAIK, this was one of the only mods which did this.  There was a missing check looking to see if the image had already been loaded

**Comment by linuxgurugamer** [comment]
CKAN is being a bit slow in updating, but you can download and install it yourself from either github or spacedock
